onnxruntime/core/providers/rocm/tunable/gemm_rocblas.h

```diff
@@ -32,6 +33,181 @@ class RocblasHandleStreamGuard {
+  return rocblas_datatype_f64_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasComputeTypeFor<BFloat16>(const BFloat16*) {
+  return rocblas_datatype_f32_r;
+}
+
+template <typename T>
+class IndexedRocBlasGemmOp
+{
+public:
+  IndexedRocBlasGemmOp()
+      : index_(0) {}
+  IndexedRocBlasGemmOp(int index)
+      : index_(index) {}
+
+  Status operator()(const GemmParams<T>* params) {
+    RocblasHandleStreamGuard guard(params->handle, params->stream);
+    return ROCBLAS_CALL(
+      rocblas_gemm_ex(
+        params->handle,
+        params->opb == BlasOp::N ? rocblas_operation_none : rocblas_operation_transpose,
+        params->opa == BlasOp::N ? rocblas_operation_none : rocblas_operation_transpose,
+        params->n, params->m, params->k,
+        &(params->alpha),
+        params->b, RocBlasDataTypeFor(params->b), params->ldb,
+        params->a, RocBlasDataTypeFor(params->a), params->lda,
+        &(params->beta),
+        params->c, RocBlasDataTypeFor(params->c), params->ldc,
+        params->c, RocBlasDataTypeFor(params->c), params->ldc,
+        RocBlasComputeTypeFor(params->a),
+        rocblas_gemm_algo_standard,
+        index_,
+        rocblas_gemm_flags_none
+      )
+    );
+  }
+
+  Status IsSupported(const GemmParams<T>*) {
+    return Status::OK();
+  }
+
+private:
+  int index_;
+};
+
+template <typename T>
+class RocBlasGemmTunableOp : public tunable::TunableOp<GemmParams<T>>
+{
+public:
+  RocBlasGemmTunableOp() {
+    // Ensure that the default implementation is always present
+    this->ops_.emplace_back(IndexedRocBlasGemmOp<T>{0});
+  }
+
+  Status IsSupported(const GemmParams<T>* params) {
+    ORT_UNUSED_PARAMETER(params);
+    return Status::OK();
+  }
+
+protected:
+  virtual int FindFastest(const GemmParams<T>* params) override {
+    auto solution_indices = this->GetSolutions(params);
+    std::vector<Op<GemmParams<T>>> candidates;
+    for (int solution_idx : solution_indices) {
+      candidates.emplace_back(IndexedRocBlasGemmOp<T>{solution_idx});
+    }
+
+    auto id = this->FindFastestImpl(params, candidates);
+    // memoize the result
+    this->ops_.emplace_back(std::move(candidates[id]));
+    return this->ops_.size() - 1;
+  }
+
+private:
+  std::vector<int> GetSolutions(const GemmParams<T>* params) {
+    int num_solutions = 0;
+    // Get the number of candidate solutions
+    ROCBLAS_CALL_THROW(rocblas_gemm_ex_get_solutions(
+      params->handle,
+      params->opb == BlasOp::N ? rocblas_operation_none : rocblas_operation_transpose,
+      params->opa == BlasOp::N ? rocblas_operation_none : rocblas_operation_transpose,
+      params->n, params->m, params->k,
+      &(params->alpha),
+      params->b, RocBlasDataTypeFor(params->b), params->ldb,
+      params->a, RocBlasDataTypeFor(params->a), params->lda,
+      &(params->beta),
+      params->c, RocBlasDataTypeFor(params->c), params->ldc,
+      params->c, RocBlasDataTypeFor(params->c), params->ldc,
```

Comment:
Why this line is repeated here? Is it some weirdness in their API?

Reply:
This is the standard rocBLAS GEMM API: after specifying the pointer, type and ld for the C tensor, you need to specify the same info for the destination tensor (they refer to it as D in the docs). The D tensor can be the same as the C tensor, which is why the line is repeated.